packages/fx-core/src/core/middleware/consolidateLocalRemote.ts

```diff
@@ -0,0 +1,312 @@
+      await fs.writeFile(manifestTemplatePath, JSON.stringify(manifest, null, 4));
+      sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateAddSPFXManifest);
+    } else {
+      const remoteManifestFile = path.join(
+        inputs.projectPath as string,
+        "templates",
+        "appPackage",
+        "manifest.remote.template.json"
+      );
+      const remoteManifestExist = await fs.pathExists(remoteManifestFile);
+      if (remoteManifestExist) {
+        sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateCopyAzureManifestStart);
+        const manifestTemplatePath = await getManifestTemplatePath(inputs.projectPath as string);
+        await fs.copyFile(remoteManifestFile, manifestTemplatePath);
+        sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateCopyAzureManifest);
+      }
+    }
+    fileList.push(
+      path.join(inputs.projectPath as string, "templates", "appPackage", "template.manifest.json")
+    );
+  } catch (e) {
+    for (const item of fileList) {
+      await fs.remove(item);
+    }
+    throw e;
+  }
+
+  // move old configs
+  const backupPath = path.join(inputs.projectPath as string, backupFolder);
+  sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateBackupConfigStart);
+  const localSettingsFile = path.join(
+    inputs.projectPath as string,
+    ".fx",
+    "configs",
+    "localSettings.json"
+  );
+  if (await fs.pathExists(localSettingsFile)) {
+    await fs.move(
+      localSettingsFile,
+      path.join(backupPath, ".fx", "configs", "localSettings.json"),
+      { overwrite: true }
+    );
+  }
+  const localManifestFile = path.join(
+    inputs.projectPath as string,
+    "templates",
+    "appPackage",
+    "manifest.local.template.json"
+  );
+  if (await fs.pathExists(localManifestFile)) {
+    await fs.move(
+      localManifestFile,
+      path.join(backupPath, "templates", "appPackage", "manifest.local.template.json"),
+      { overwrite: true }
+    );
+  }
+  const remoteManifestFile = path.join(
+    inputs.projectPath as string,
+    "templates",
+    "appPackage",
+    "manifest.remote.template.json"
+  );
+  if (await fs.pathExists(remoteManifestFile)) {
+    await fs.move(
+      remoteManifestFile,
+      path.join(backupPath, "templates", "appPackage", "manifest.remote.template.json"),
+      { overwrite: true }
+    );
+  }
+
+  sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateBackupConfig);
+  sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateUpgrade);
+
+  postConsolidate(inputs.projectPath as string, ctx, inputs, backupFolder);
+  return true;
+}
+
+function checkMethod(ctx: CoreHookContext): boolean {
+  const methods: Set<string> = new Set(["getProjectConfig", "checkPermission"]);
+  if (ctx.method && methods.has(ctx.method) && userCancelFlag) return false;
+  userCancelFlag = ctx.method != undefined && methods.has(ctx.method);
```

Comment:
A little bit confused about the `userCancelFlag` variable, seems the value is not from user cancel operation but the fx-core methods checking?

Reply:
When opening new fx project, it will enter ```consolidateLocalRemote``` logic multi times so if user cancel we need to save it temporarily. When provision or local debug, we need to ask user whether to upgrade every time. Check method is used to separate different enter point.

Reply:
> ... if user cancel we need to save it temporarily.

Seems there's no logic to save the 'user cancel' info according to L66-L72?

Reply:
Save this not when the user cancels but when the question is first asked during VsCode opens ```teamsfx``` project. May be ```user cancel``` is not a good name..